ddm-admin-client/build.rs

```diff
@@ -0,0 +1,99 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+// Copyright 2022 Oxide Computer Company
+
+use anyhow::bail;
+use anyhow::Context;
+use anyhow::Result;
+use quote::quote;
+use serde::Deserialize;
+use std::collections::BTreeMap;
+use std::env;
+use std::fs;
+use std::path::Path;
+
+// TODO-cleanup This is a copy/paste of (a subset of) omicron_package::Config.
+// We ought to depend on it directly, but that creates a circular dependency
+// between crates:
+//
+// us -> omicron_package -> omicron_sled_agent -> us
+//
+// omicron_package only depends on sled-agent for zone management; maybe that
+// could be broken out into a separate crate?
+#[derive(Deserialize, Debug)]
+#[serde(tag = "type", rename_all = "lowercase")]
+pub enum ExternalPackageSource {
+    Prebuilt { repo: String, commit: String, sha256: String },
+    Manual,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct ExternalPackage {
+    pub source: ExternalPackageSource,
+}
+
+#[derive(Deserialize, Debug)]
+pub struct Config {
+    #[serde(default, rename = "external_package")]
+    pub external_packages: BTreeMap<String, ExternalPackage>,
+}
+
+fn main() -> Result<()> {
+    // Find the current maghemite repo commit from our package manifest.
+    let manifest = fs::read("../package-manifest.toml")
+        .context("failed to read ../package-manifest.toml")?;
+    println!("cargo:rerun-if-changed=../package-manifest.toml");
+
+    let config: Config = toml::de::from_slice(&manifest)
+        .context("failed to parse ../package-manifest.toml")?;
+    let maghemite = config
+        .external_packages
+        .get("maghemite")
+        .context("missing maghemite package in ../package-manifest.toml")?;
+    let commit = match &maghemite.source {
+        ExternalPackageSource::Manual => {
+            bail!("maghemite external package must have type `prebuilt`")
+        }
+        ExternalPackageSource::Prebuilt { commit, .. } => commit,
+    };
+
+    let url = format!("https://buildomat.eng.oxide.computer/public/file/oxidecomputer/maghemite/openapi/{commit}/ddm-admin.json");
```

Comment:
My *only* hesitation here is that a clean build, even after installing pre-requisites, will now require internet access.

I'm kinda on the fence about this.

- On the one hand, the way that you have implemented this is *convenient*. Simply re-running the build gets the latest API.
- On the other hand, network requests in build.rs files is kinda sketchy?

I dunno. We can land-this as-is, but I do want to make sure we're aware of the cost of this kind of build script.

Reply:
@smklein How does 37b06f4e look?

1. No more internet access in `build.rs`
2. There is some manual sync required when updating the maghemite commit (both package-manifest.toml and tools/maghemite_openapi_version must be updated), but only updating package-manifest.toml will be caught when trying to build ddm-admin-client.

Reply:
That approach looks great; thanks for taking the time to fix.

I think the "internet access when installing pre-requisites" step is totally expected, so this seems appropriate. Thanks for making the download script idempotent too.